### PR TITLE
Support seq len equal to 0 in sequence ops

### PR DESCRIPTION
**Pull request by kuke**
test=develop

### CODE REVIEW

File: paddle/fluid/operators/math/context_project.h [hunk 1]
```diff
@@ -104,6 +104,10 @@ class ContextProjectFunctor {
     sequence_width = in.dims()[1];
 
     for (int i = 0; i < static_cast<int>(lod_level_0.size()) - 1; ++i) {
+      if (i < static_cast<int>(lod_level_0.size()) - 2 &&
```
Comment (phlrain) [inline paddle/fluid/operators/math/context_project.h]: 根据上面的for 循环，i < static_cast<int>(lod_level_0.size()) - 2  好像一直成立

File: paddle/fluid/operators/math/context_project.h [hunk 2]
```diff
@@ -133,6 +137,10 @@ class ContextProjectFunctor {
     }
     if (padding_trainable) {
       for (int i = 0; i < static_cast<int>(lod_level_0.size()) - 1; ++i) {
+        if (i < static_cast<int>(lod_level_0.size()) - 2 &&
```
Comment (phlrain) [inline paddle/fluid/operators/math/context_project.h]: 同上

File: paddle/fluid/operators/math/context_project.h [hunk 3]
```diff
@@ -215,6 +223,10 @@ class ContextProjectGradFunctor {
 
     if (input_grad) {
       for (int i = 0; i < static_cast<int>(lod_level_0.size()) - 1; ++i) {
+        if (i < static_cast<int>(lod_level_0.size()) - 2 &&
```
Comment (phlrain) [inline paddle/fluid/operators/math/context_project.h]: 同上

File: paddle/fluid/operators/math/context_project.h [hunk 4]
```diff
@@ -247,6 +259,10 @@ class ContextProjectGradFunctor {
     if (pad_grad) {
       if (padding_trainable) {
         for (int i = 0; i < static_cast<int>(lod_level_0.size()) - 1; ++i) {
+          if (i < static_cast<int>(lod_level_0.size()) - 2 &&
+              lod_level_0[i] == lod_level_0[i + 1])
```
Comment (phlrain) [inline paddle/fluid/operators/math/context_project.h]: 同上

File: paddle/fluid/operators/sequence_ops/sequence_enumerate_op.h [hunk 5]
```diff
@@ -49,6 +49,8 @@ class SequenceEnumerateKernel : public framework::OpKernel<T> {
     for (size_t i = 0; i < lod0.size() - 1; ++i) {
       int start = lod0[i];
       int end = lod0[i + 1];
+      if (start >= end) continue;
```
Comment (phlrain) [inline paddle/fluid/operators/sequence_ops/sequence_enumerate_op.h]: 按照约定, start 不会大于 end吧

**Review by qingqing01 — COMMENTED** [review]
LGTM.

File: paddle/fluid/operators/sequence_ops/sequence_slice_op.h [hunk 6]
```diff
@@ -77,7 +77,7 @@ class SequenceSliceOpKernel : public framework::OpKernel<T> {
     for (size_t i = 0; i < n; ++i) {
       PADDLE_ENFORCE_LE(0, offset_data[i],
                         "The offset[%d] must greater than zero.", i);
-      PADDLE_ENFORCE_LT(0, length_data[i],
+      PADDLE_ENFORCE_LE(0, length_data[i],
                         "The length[%d] must greater than zero.", i);
```
Comment (phlrain) [inline paddle/fluid/operators/sequence_ops/sequence_slice_op.h]: greater than or same with

File: python/paddle/fluid/tests/unittests/test_sequence_concat.py [hunk 7]
```diff
@@ -41,5 +46,19 @@ def test_dx(self):
...
+        self.out_lod = [22, 8]
+
+
+class TestSequenceConcatCase3(TestSequenceConcat):
+    def setLoD(self):
+        self.lod1 = [10, 0]
+        self.lod2 = [20, 0]
+        self.out_lod = [30, 0]
+
+
```
Comment (phlrain) [inline python/paddle/fluid/tests/unittests/test_sequence_concat.py:62]: 添两个单测 lod1 = [0, 10],  lod2=[ 20, 10]
lod1 = [ 0, 10],   lod2 = [0, 10]

File: python/paddle/fluid/tests/unittests/test_sequence_slice_op.py [hunk 8]
```diff
@@ -59,5 +59,13 @@ def test_check_grad(self):
         self.check_grad(['X'], 'Out')
 
 
+class TestSequenceSliceOpSeqlen0(TestSequenceSliceOp):
+    def init_test_case(self):
+        self.x_dim = (100, 3, 2)
+        self.x_lod = [[20, 30, 0, 30, 20]]
+        self.offset = [[1], [2], [0], [4], [5]]
+        self.length = [[10], [8], [0], [4], [2]]
+
```
Comment (phlrain) [inline python/paddle/fluid/tests/unittests/test_sequence_slice_op.py:68]: 添加几个测试的用例吧
x_lod = [ [ 0, 20, 30, 20, 30]]
x_lod =[ [ 20, 30, 20, 30, 0]]
x_lod = [ [ 0, 0, 0, 50, 50]]
x_lod = [[50, 50, 0, 0, 0]]

File: python/paddle/fluid/tests/unittests/test_sequence_enumerate_op.py [hunk 9]
```diff
@@ -101,5 +101,16 @@ def init_test_case(self):
...
+    def init_test_case(self):
+        self.in_seq = np.random.randint(0, 10, (30, 1)).astype("int32")
+        self.lod = [[9, 4, 11, 0, 6]]
+        self.win_size = 5
+        self.pad_value = 5
+        out_seq = sequence_enumerate(self.in_seq, self.lod, self.win_size,
+                                     self.pad_value)
+        self.out_seq = np.array(out_seq).astype("int32")
+
+
```
Comment (phlrain) [inline python/paddle/fluid/tests/unittests/test_sequence_enumerate_op.py:114]: LOD 第1个，最后一个位置，为0的测试用例

File: paddle/fluid/operators/math/context_project.h (hunk 1 above)
Comment (kuke) [reply paddle/fluid/operators/math/context_project.h]: Done. This is a bug

File: paddle/fluid/operators/math/context_project.h (hunk 2 above)
Comment (kuke) [reply paddle/fluid/operators/math/context_project.h]: Done

File: paddle/fluid/operators/math/context_project.h (hunk 3 above)
Comment (kuke) [reply paddle/fluid/operators/math/context_project.h]: Done

File: paddle/fluid/operators/math/context_project.h (hunk 4 above)
Comment (kuke) [reply paddle/fluid/operators/math/context_project.h]: Done

File: paddle/fluid/operators/sequence_ops/sequence_enumerate_op.h (hunk 5 above)
Comment (kuke) [reply paddle/fluid/operators/sequence_ops/sequence_enumerate_op.h]: Changed

**Review by phlrain — DISMISSED** [review]

**Review by phlrain — DISMISSED** [review]

**Review by junjun315 — DISMISSED** [review]
LGTM

**Review by sneaxiy — APPROVED** [review]
LGTM.

**Pull request merged**